### PR TITLE
add a metirc of ideal state size after gzip

### PR DESCRIPTION
**Pull request by yupeng9**
Today controller emits a metric `pinot.controller.idealstateZnodeSize`. The znode size is useful to monitor znode size is less than 1MB, otherwise Helix cannot process the znode. However, the reported size is before the gzip compression, not the actual size stored in zk.  Add a new metric of the znode size after gzip. Note that the difference of the znode size before and after gzip could be significant (e.g. >100X~1000X).

**<code>release-notes</code>**
Add a new metric `pinot.controller.idealstateZnodeGzipSize` to emit the idealstate znode size after gzip compression.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java [hunk 1]
```diff
@@ -145,7 +148,9 @@ private void updateSegmentMetrics(String tableNameWithType, Context context) {
     }
 
     _controllerMetrics
-        .setValueOfTableGauge(tableNameWithType, ControllerGauge.IDEALSTATE_ZNODE_SIZE, idealState.toString().length());
+        .setValueOfTableGauge(tableNameWithType, ControllerGauge.IDEALSTATE_ZNODE_SIZE,  idealState.toString().length());
```
Comment (xiangfu0) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:151]: extra space between 
```
ControllerGauge.IDEALSTATE_ZNODE_SIZE,  idealState.toString().length());
```

**Review by xiangfu0 — APPROVED** [review]

**Comment by mcvsubbu** [comment]
This could explode the number of metrics we have in our cluster that has 3.5k+ tables. Why do we need this?

**Review by Jackie-Jiang — COMMENTED** [review]
It is not always gzipped. I think we should just change `IDEALSTATE_ZNODE_SIZE` to the one you added. Metric on string length provides very less value IMO

**Comment by yupeng9** [comment]
> It is not always gzipped. I think we should just change `IDEALSTATE_ZNODE_SIZE` to the one you added. Metric on string length provides very less value IMO

It'll be back-incompat break. @xiangfu0 what do you think?

**Comment by xiangfu0** [comment]
> This could explode the number of metrics we have in our cluster that has 3.5k+ tables. Why do we need this?

The major reason is that we want to alert when there are many segments in a table, which will hit the znode size limit which is 1MB.

**Comment by xiangfu0** [comment]
> > It is not always gzipped. I think we should just change `IDEALSTATE_ZNODE_SIZE` to the one you added. Metric on string length provides very less value IMO
> 
> It'll be back-incompat break. @xiangfu0 what do you think?

I think we want to know both numbers. for znode size, when is the time it will get gzipped and how can we set an alert on this? Say we want to get notified on 750kb for warning and 850kb for critical

**Comment by Jackie-Jiang** [comment]
How about renaming it to `IDEALSTATE_ZNODE_BYTE_SIZE`?
I'm not seeing much value on tracking `ZNRecord.toString().length` because that's not how the serializer serialize the record, but I'm okay keeping both of them. Maybe better to change it to reflect the accurate byte size without compression, but not in the scope of this PR.

**Comment by yupeng9** [comment]
`IDEALSTATE_ZNODE_BYTE_SIZE ` is ok to me. @xiangfu0 what do you think?

**Comment by xiangfu0** [comment]
> `IDEALSTATE_ZNODE_BYTE_SIZE ` is ok to me. @xiangfu0 what do you think?

sounds good

**Pull request merged**